### PR TITLE
[MAINTENANCE] Fixing checkpoint types

### PR DESCRIPTION
**Pull request by anthonyburdi**
Changes proposed in this pull request:
- Some progress toward fixing type errors in checkpoint.py and related files.
- Remove util method get_datetime_string_from_strftime_format and inline strftime instead.

### Definition of Done
Please delete options that are not relevant.

- [x] My code follows the Great Expectations [style guide](https://docs.greatexpectations.io/docs/contributing/style_guides/code_style)
- [x] I have performed a [self-review](https://docs.greatexpectations.io/docs/contributing/contributing_checklist) of my own code
- [x] I have run any local integration tests and made sure that nothing is broken.

Thank you for submitting!

### CODE REVIEW

File: great_expectations/checkpoint/checkpoint.py [hunk 1]
```diff
@@ -127,6 +127,7 @@ def run(
         run_time = run_time or datetime.datetime.now()
         runtime_configuration = runtime_configuration or {}
         result_format = result_format or runtime_configuration.get("result_format")
+        assert isinstance(result_format, (NoneType, str, dict))
```
Comment (cdkini) [inline great_expectations/checkpoint/checkpoint.py]: Should we add an error message here?

**Review by cdkini — APPROVED** [review]

File: great_expectations/checkpoint/checkpoint.py (hunk 1 above)
Comment (anthonyburdi) [reply great_expectations/checkpoint/checkpoint.py]: Done!

File: great_expectations/checkpoint/checkpoint.py [hunk 2]
```diff
@@ -127,6 +127,9 @@ def run(
         run_time = run_time or datetime.datetime.now()
         runtime_configuration = runtime_configuration or {}
         result_format = result_format or runtime_configuration.get("result_format")
+        assert isinstance(
+            result_format, (NoneType, str, dict)
```
Comment (Kilo59) [inline great_expectations/checkpoint/checkpoint.py]:
```suggestion
            result_format, (type(None), str, dict)
```

Comment (anthonyburdi) [reply great_expectations/checkpoint/checkpoint.py]: Done!

File: great_expectations/checkpoint/checkpoint.py [hunk 3]
```diff
@@ -127,6 +126,9 @@ def run(
         run_time = run_time or datetime.datetime.now()
         runtime_configuration = runtime_configuration or {}
         result_format = result_format or runtime_configuration.get("result_format")
+        assert isinstance(
+            result_format, (type(None), str, dict)
+        ), "result_format should either be a string or dict if supplied, or None if not supplied."
```
Comment (Kilo59) [inline great_expectations/checkpoint/checkpoint.py]:
```suggestion
        _result_fmt_types = (type(None), str, dict)
        assert isinstance(
            result_format, result_format_types
        ), f"result_format should be of type - {' '.join(_result_fmt_types)}"

```

Could do something like this 🤷

File: great_expectations/core/util.py [hunk 4]
```diff
@@ -429,7 +429,7 @@ def substitute_all_strftime_format_strings(
 
 def parse_string_to_datetime(
     datetime_string: str, datetime_format_string: Optional[str] = None
-) -> datetime.date:
+) -> datetime.datetime:
```
Comment (anthonyburdi) [inline great_expectations/core/util.py:432]: This method is always returning a datetime

File: great_expectations/core/expectation_suite.py [hunk 5]
```diff
@@ -128,12 +129,22 @@ def add_citation(
...
+            citation_date_obj = citation_date
+        else:
+            raise GreatExpectationsTypeError(
+                f"citation_date should be of type - {' '.join(str(t) for t in _citation_date_types)}"
+            )
 
-        citation_date = citation_date or datetime.datetime.now(datetime.timezone.utc)
         citation: Dict[str, Any] = {
-            "citation_date": citation_date.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
+            "citation_date": citation_date_obj.strftime("%Y-%m-%dT%H:%M:%S.%fZ"),
```
Comment (anthonyburdi) [inline great_expectations/core/expectation_suite.py]: I thought this was more clear, but open to feedback.

**Pull request merged**